File: tests/algorithms/transpiler/FS/data_structures/stacks/prefix_evaluation.fs
// Generated 2025-08-08 11:10 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _substring (s:string) (start:int) (finish:int) =
    let len = String.length s
    let mutable st = if start < 0 then len + start else start
    let mutable en = if finish < 0 then len + finish else finish
    if st < 0 then st <- 0
    if st > len then st <- len
    if en > len then en <- len
    if st > en then st <- en
    s.Substring(st, en - st)

let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec split_custom (s: string) (sep: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable s = s
    let mutable sep = sep
    try
        let mutable res: string array = [||]
        let mutable current: string = ""
        let mutable i: int = 0
        while i < (String.length (s)) do
            let ch: string = _substring s i (i + 1)
            if ch = sep then
                res <- Array.append res [|current|]
                current <- ""
            else
                current <- current + ch
            i <- i + 1
        res <- Array.append res [|current|]
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec tokenize (s: string) =
    let mutable __ret : string array = Unchecked.defaultof<string array>
    let mutable s = s
    try
        let parts: string array = split_custom (s) (" ")
        let mutable res: string array = [||]
        let mutable i: int = 0
        while i < (Seq.length (parts)) do
            let p: string = _idx parts (i)
            if p <> "" then
                res <- Array.append res [|p|]
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec is_digit (ch: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable ch = ch
    try
        __ret <- (ch >= "0") && (ch <= "9")
        raise Return
        __ret
    with
        | Return -> __ret
let rec is_operand (token: string) =
    let mutable __ret : bool = Unchecked.defaultof<bool>
    let mutable token = token
    try
        if token = "" then
            __ret <- false
            raise Return
        let mutable i: int = 0
        while i < (String.length (token)) do
            let ch: string = _substring token i (i + 1)
            if not (is_digit (ch)) then
                __ret <- false
                raise Return
            i <- i + 1
        __ret <- true
        raise Return
        __ret
    with
        | Return -> __ret
let rec to_int (token: string) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable token = token
    try
        let mutable res: int = 0
        let mutable i: int = 0
        while i < (String.length (token)) do
            res <- (res * 10) + (int (_substring token i (i + 1)))
            i <- i + 1
        __ret <- res
        raise Return
        __ret
    with
        | Return -> __ret
let rec apply_op (op: string) (a: float) (b: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable op = op
    let mutable a = a
    let mutable b = b
    try
        if op = "+" then
            __ret <- a + b
            raise Return
        if op = "-" then
            __ret <- a - b
            raise Return
        if op = "*" then
            __ret <- a * b
            raise Return
        if op = "/" then
            __ret <- a / b
            raise Return
        __ret <- 0.0
        raise Return
        __ret
    with
        | Return -> __ret
let rec evaluate (expression: string) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable expression = expression
    try
        let tokens: string array = tokenize (expression)
        let mutable stack: float array = [||]
        let mutable i: int = (Seq.length (tokens)) - 1
        while i >= 0 do
            let token: string = _idx tokens (i)
            if token <> "" then
                if is_operand (token) then
                    stack <- Array.append stack [|(float (to_int (token)))|]
                else
                    let o1: float = _idx stack ((Seq.length (stack)) - 1)
                    let o2: float = _idx stack ((Seq.length (stack)) - 2)
                    stack <- Array.sub stack 0 (((Seq.length (stack)) - 2) - 0)
                    let mutable res: float = apply_op (token) (o1) (o2)
                    stack <- Array.append stack [|res|]
            i <- i - 1
        __ret <- _idx stack (0)
        raise Return
        __ret
    with
        | Return -> __ret
let rec eval_rec (tokens: string array) (pos: int) =
    let mutable __ret : float array = Unchecked.defaultof<float array>
    let mutable tokens = tokens
    let mutable pos = pos
    try
        let token: string = _idx tokens (pos)
        let next: int = pos + 1
        if is_operand (token) then
            __ret <- unbox<float array> [|float (to_int (token)); float next|]
            raise Return
        let left: float array = eval_rec (tokens) (next)
        let a: float = _idx left (0)
        let p1: int = int (_idx left (1))
        let right: float array = eval_rec (tokens) (p1)
        let b: float = _idx right (0)
        let p2: float = _idx right (1)
        __ret <- unbox<float array> [|apply_op (token) (a) (b); p2|]
        raise Return
        __ret
    with
        | Return -> __ret
let rec evaluate_recursive (expression: string) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable expression = expression
    try
        let tokens: string array = tokenize (expression)
        let mutable res: float array = eval_rec (tokens) (0)
        __ret <- _idx res (0)
        raise Return
        __ret
    with
        | Return -> __ret
let test_expression: string = "+ 9 * 2 6"
printfn "%s" (_str (evaluate (test_expression)))
let test_expression2: string = "/ * 10 2 + 4 1 "
printfn "%s" (_str (evaluate (test_expression2)))
let test_expression3: string = "+ * 2 3 / 8 4"
printfn "%s" (_str (evaluate_recursive (test_expression3)))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/PHP/data_structures/stacks/prefix_evaluation.php
<?php
error_reporting(E_ALL & ~E_DEPRECATED);
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function split_custom($s, $sep) {
  global $test_expression, $test_expression2, $test_expression3;
  $res = [];
  $current = '';
  $i = 0;
  while ($i < strlen($s)) {
  $ch = substr($s, $i, $i + 1 - $i);
  if ($ch == $sep) {
  $res = _append($res, $current);
  $current = '';
} else {
  $current = $current . $ch;
}
  $i = $i + 1;
};
  $res = _append($res, $current);
  return $res;
};
  function tokenize($s) {
  global $test_expression, $test_expression2, $test_expression3;
  $parts = split_custom($s, ' ');
  $res = [];
  $i = 0;
  while ($i < count($parts)) {
  $p = $parts[$i];
  if ($p != '') {
  $res = _append($res, $p);
}
  $i = $i + 1;
};
  return $res;
};
  function is_digit($ch) {
  global $test_expression, $test_expression2, $test_expression3;
  return $ch >= '0' && $ch <= '9';
};
  function is_operand($token) {
  global $test_expression, $test_expression2, $test_expression3;
  if ($token == '') {
  return false;
}
  $i = 0;
  while ($i < strlen($token)) {
  $ch = substr($token, $i, $i + 1 - $i);
  if (!is_digit($ch)) {
  return false;
}
  $i = $i + 1;
};
  return true;
};
  function to_int($token) {
  global $test_expression, $test_expression2, $test_expression3;
  $res = 0;
  $i = 0;
  while ($i < strlen($token)) {
  $res = $res * 10 + ((ctype_digit($token[$i]) ? intval($token[$i]) : ord($token[$i])));
  $i = $i + 1;
};
  return $res;
};
  function apply_op($op, $a, $b) {
  global $test_expression, $test_expression2, $test_expression3;
  if ($op == '+') {
  return $a + $b;
}
  if ($op == '-') {
  return $a - $b;
}
  if ($op == '*') {
  return $a * $b;
}
  if ($op == '/') {
  return $a / $b;
}
  return 0.0;
};
  function evaluate($expression) {
  global $test_expression, $test_expression2, $test_expression3;
  $tokens = tokenize($expression);
  $stack = [];
  $i = count($tokens) - 1;
  while ($i >= 0) {
  $token = $tokens[$i];
  if ($token != '') {
  if (is_operand($token)) {
  $stack = _append($stack, (floatval(to_int($token))));
} else {
  $o1 = $stack[count($stack) - 1];
  $o2 = $stack[count($stack) - 2];
  $stack = array_slice($stack, 0, count($stack) - 2);
  $res = apply_op($token, $o1, $o2);
  $stack = _append($stack, $res);
};
}
  $i = $i - 1;
};
  return $stack[0];
};
  function eval_rec($tokens, $pos) {
  global $test_expression, $test_expression2, $test_expression3;
  $token = $tokens[$pos];
  $next = $pos + 1;
  if (is_operand($token)) {
  return [(floatval(to_int($token))), (floatval($next))];
}
  $left = eval_rec($tokens, $next);
  $a = $left[0];
  $p1 = intval($left[1]);
  $right = eval_rec($tokens, $p1);
  $b = $right[0];
  $p2 = $right[1];
  return [apply_op($token, $a, $b), $p2];
};
  function evaluate_recursive($expression) {
  global $test_expression, $test_expression2, $test_expression3;
  $tokens = tokenize($expression);
  $res = eval_rec($tokens, 0);
  return $res[0];
};
  $test_expression = '+ 9 * 2 6';
  echo rtrim(_str(evaluate($test_expression))), PHP_EOL;
  $test_expression2 = '/ * 10 2 + 4 1 ';
  echo rtrim(_str(evaluate($test_expression2))), PHP_EOL;
  $test_expression3 = '+ * 2 3 / 8 4';
  echo rtrim(_str(evaluate_recursive($test_expression3))), PHP_EOL;
$__end = _now();
$__end_mem = memory_get_peak_usage(true);
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
